### PR TITLE
[Typechecker] Emit specialized diagnostic notes on automatic synthesis failure to Comparable

### PR DESCRIPTION
**Pull request by theblixguy**
Automatic synthesis of `Comparable` is only supported for `enum`s for now. So, if we have a `struct` or a `class` and a conformance failure to `Comparable`, emit a note just like we do for `Equatable` (on classes) for example, to let the user know that automatic synthesis is unsupported.

As a bonus, also emit a note when the `enum` has a raw type or if the associated values don't conform to `Comparable`.

Resolves SR-13148
rdar://problem/65116465

### CODE REVIEW

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -4400,6 +4400,15 @@ void swift::diagnoseConformanceFailure(Type T,
 
   diags.diagnose(ComplainLoc, diag::type_does_not_conform,
                  T, Proto->getDeclaredType());
+
+  // Special case: Automatic synthesis of Comparable only works with enums for
+  // now. If we are trying to conform to Comparable and failing, emit a note to
+  // let the user know about this.
+  if (!T->getEnumOrBoundGenericEnum() &&
```
Comment (theblixguy) [inline lib/Sema/TypeCheckProtocol.cpp:4407]: I have added the check here because we simply do not attempt any synthesis if `derivesProtocolRequirement` returns false. I thought about putting the check there but I didn't want the diagnostic to be a side-effect of performing a check (and `derivesProtocolRequirement` doesn't do any diagnostics anyway and is only to check if it's possible to synthesise the requirement or not).

Let me know if you think this should go elsewhere :)

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2833,9 +2833,9 @@ NOTE(missing_member_type_conformance_prevents_synthesis, none,
      "protocol %2, preventing synthesized conformance "
      "of %3 to %2",
      (unsigned, Type, Type, Type))
-NOTE(classes_automatic_protocol_synthesis,none,
-    "automatic synthesis of '%0' is not supported for classes",
-    (StringRef))
+NOTE(automatic_protocol_synthesis_unsupported,none,
+    "automatic synthesis of '%0' is not supported for %select{classes|structs}1",
+    (StringRef, unsigned))
```
Comment (xedin) [inline include/swift/AST/DiagnosticsSema.def:2838]: Please make this modification in `include/swift/AST/diagnostics/en.yaml` as well, that would help us a lot :)

Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def:2838]: Done :)

Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:2838]: Awesome, thank you!

**Comment by theblixguy** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/TypeCheckProtocol.cpp:4407]: Is there a place where compiler detect that conformance requirements need to be derived automatically? That might be a place to diagnose this...

Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp:4407]: Hmm, I’ve looked around a bit but I couldn’t find a suitable place other than this, because the code for `Comparable` synthesis doesn’t even get invoked in situations like this. 

My understanding is that the compiler resolves witnesses using three strategies - it attempts to resolve it using name lookup, if we fail to find a witness, then it tries to derive it and if we still fail to find it then it tries to find a default implementation.

In the second step (derivation), the compiler first checks if it’s possible to derive the witness (by calling `derivesProtocolRequirement`). If it’s not possible to derive it, it just bails out. 

Now, we _could_ emit the note in `DeriveConformanceComparable.cpp` but to do that would would have to move the “can I derive the `Comparable` requirement?” check from `derivesProtocolRequirement` into `deriveComparable` itself, but I am not sure if that’s the right thing to do. What do you think?

Comment (xedin) [reply lib/Sema/TypeCheckProtocol.cpp:4407]: What about `DerivedConformance::tryDiagnoseFailedDerivation`? It seems like `resolveWitnessViaLookup` has all the necessary information to determine whether it's even possible to attempt to derive requirements.

File: test/decl/protocol/special/comparable/comparable_unsupported_type.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  // expected-error@-1 {{type 'NotComparableClass' does not conform to protocol 'Comparable'}}
+  // expected-note@-2 {{automatic synthesis of 'Comparable' is not supported for classes}}
+  // expected-error@-3 {{type 'NotComparableClass' does not conform to protocol 'Equatable'}}
+  // expected-note@-4 {{automatic synthesis of 'Equatable' is not supported for classes}}
+  var value = 1
+}
+
+// Make sure we don't emit the note for enums.
+
+enum NotComparableEnum: Int, Comparable {
```
Comment (LucianoPAlmeida) [inline test/decl/protocol/special/comparable/comparable_unsupported_type.swift:21]: @theblixguy @xedin Do you think its is worth a note mentioning that enum with raw values cannot synthesize comparable? Maybe its something to throw in some educational notes explaining that ... what you think?
The question is more because if we have a diagnostic that says that comparable conformances can only be synthesized for enums and then have an enum that fails synthesize/conformance, it may be confusing(the why is failing) especially for beginners. Do you thinks this makes sense and its worth giving a bit more information in this case? Perhaps the educational notes would be good for something like this :)

File: test/decl/protocol/special/comparable/comparable_unsupported_type.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  // expected-error@-1 {{type 'NotComparableEnum' does not conform to protocol 'Comparable'}}
+  case value = 2
+}
```
Comment (varungandhi-apple) [inline test/decl/protocol/special/comparable/comparable_unsupported_type.swift:24]: Not part of this PR but I find it puzzling (and I didn't realize this before) that such enums do not support automatic synthesis of `Comparable`. 🤔

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp:4407]: You're right, looks like I didn't notice that function at all 🤦 I have moved the check there now!

File: test/decl/protocol/special/comparable/comparable_unsupported_type.swift (hunk 3 above)
Comment (theblixguy) [reply test/decl/protocol/special/comparable/comparable_unsupported_type.swift:21]: I think it makes sense, we could emit a different note to tell the user that, something like `raw values prevent automatic 'Comparable' synthesis` or `enum 'Foo' has raw values, preventing synthesized conformance to `Comparable`. WDYT?

Comment (theblixguy) [reply test/decl/protocol/special/comparable/comparable_unsupported_type.swift:21]: We could also add another note for payload cases where the payload does not conform to `Comparable`. I believe we already have a diagnostic note for it, just needs to be applied in this case as well.

Comment (LucianoPAlmeida) [reply test/decl/protocol/special/comparable/comparable_unsupported_type.swift:21]: > I think it makes sense, we could emit a different note to tell the user that, something like `raw values prevent automatic 'Comparable' synthesis` or `enum 'Foo' has raw values, preventing synthesized conformance to 'Comparable'`. WDYT?

To me, the former seems good :)  

> We could also add another note for payload cases where the payload does not conform to Comparable. I believe we already have a diagnostic note for it, just needs to be applied in this case as well.

That would be great! Now that you mention this I'm thinking even more that beyond the note this also could benefit of having an educational note explaining the rules around comparable synthesis, not here in this patch, but something to think about adding in the future :)

Comment (theblixguy) [reply test/decl/protocol/special/comparable/comparable_unsupported_type.swift:21]: Alright, I have added the notes. Let me know what you think :) And yeah it would be good to add some educational notes in a follow-up!

**Comment by theblixguy** [comment]
The last commit is NFC - I did some refactoring as I wanted to reuse some code for the notes and avoid a lot of duplication.

@swift-ci please smoke test

File: test/decl/protocol/special/comparable/comparable_unsupported_type.swift (hunk 3 above)
Comment (LucianoPAlmeida) [reply test/decl/protocol/special/comparable/comparable_unsupported_type.swift]: Right! The note looks good to me :)

**Review by xedin — APPROVED** [review]
LGTM! But please revert all of the formatting changes to the unrelated parts of code before merging.

**Comment by theblixguy** [comment]
Sorry about that. I am actually gonna create a separate PR for the refactoring, so I am gonna revert that commit for now.

@swift-ci please smoke test

**Comment by xedin** [comment]
Sounds good!

**Comment by xedin** [comment]
@theblixguy Could you do me a favor and wait until https://github.com/apple/swift/pull/32828 lands? I have moved `en.yaml` to `diagnostics/` instead of `include/swift/AST/diagnostics`...

**Comment by theblixguy** [comment]
Sure no worries!

**Comment by xedin** [comment]
Thank you!

File: include/swift/AST/diagnostics/en.yaml [hunk 5]
```diff
@@ -6701,9 +6701,13 @@
...
 
-- id: classes_automatic_protocol_synthesis
+- id: automatic_protocol_synthesis_unsupported
   msg: >-
-    automatic synthesis of '%0' is not supported for classes
+    automatic synthesis of '%0' is not supported for %select{classes|structs}1
+
+- id: comparable_synthesis_raw_value_not_allowed
+  msg: >-
+    "enum declares raw type %0, preventing synthesized conformance of %1 to %2"
```
Comment (HassanElDesouky) [inline include/swift/AST/diagnostics/en.yaml:6710]: I don't think adding double quotations for multi-line YAML strings would work. Would be great if you can just remove the quotation.

Comment (xedin) [reply include/swift/AST/diagnostics/en.yaml:6710]: Good point. Either remove `>-` or quotations :)

Comment (HassanElDesouky) [reply include/swift/AST/diagnostics/en.yaml:6710]: Yes either way would work. But I think if all messages had the same format i.e. `>-` it will look better (because all of the messages has the same style), of course, it's not that big of a deal. :)

**Comment by xedin** [comment]
@theblixguy Localization changes went in so you can rebase and merge this PR, thank you for waiting!

**Comment by theblixguy** [comment]
@swift-ci please smoke test
